single_kernel_mongo/state/charm_state.py

```diff
@@ -280,10 +365,62 @@ def db_initialised(self) -> bool:
     def db_initialised(self, other: bool):
         self.app_peer_data.db_initialised = other
 
+    @property
+    def unit_workload_container_versions(self) -> dict[str, str]:
+        """{Unit name: unique identifier for unit's workload container version}.
+
+        If and only if this version changes, the workload will restart (during upgrade or
+        rollback).
+
+        On Kubernetes, the workload & charm are upgraded together
+        On machines, the charm is upgraded before the workload
+
+        VM: container version is the snap revision.
+        K8S: container version is the stateful set hash.
+
+        This identifier should be comparable to `_app_workload_container_version` to determine if
+        the unit & app are the same workload container version.
+        """
+        if self.substrate == Substrates.K8S:
+            return self.k8s_manager.list_revisions()
+        return {
+            unit.name: unit.snap_revision
+            for unit in self.units_upgrade_peer_data
+            if unit.snap_revision
+        }
+
+    @property
+    def unit_workload_container_version(self) -> str | None:
+        """Installed snap revision for this unit."""
+        if self.substrate == Substrates.K8S:
+            return self.k8s_manager.list_revisions().get(self.model.unit.name)
+        return self.unit_upgrade_peer_data.snap_revision
+
+    @unit_workload_container_version.setter
+    def unit_workload_container_version(self, value: str):
+        if self.substrate == Substrates.VM:
+            self.unit_upgrade_peer_data.snap_revision = value
+
+    @property
+    def app_workload_container_version(self) -> str:
+        """Unique identifier for the app's workload container version.
+
+        This should match the workload version in the current Juju app charm version.
+
+        This identifier should be comparable to `_unit_workload_container_versions` to determine if
+        the app & unit are the same workload container version.
+        """
+        if self.substrate == Substrates.K8S:
+            return self.k8s_manager.get_revision()
+        return SNAP.revision
+
     @property
     def upgrade_in_progress(self) -> bool:
         """Is the charm in upgrade?"""
-        return False
+        app_version = self.app_workload_container_version
+        unit_versions = self.unit_workload_container_versions
+        logger.debug(f"{app_version} {unit_versions}")
```

Comment:
I think this log can be removed

Reply:
It might be better to rework it to something better and keep it so it's easier to understand if some upgrades are stuck.
I'll rework that.